<Mercs and Missiles

tthreeoh opened the pull request:
I wanted to show missiles and started by showing the mercs.

Missiles are turned off by default in settings.
There are two types of missiles, Player and Enemy, and both can be turned on/off separately.

As of right now I have 13 types of missiles each with a custom color in settings.

missiles.ini is used to process what missiles are included, ignored, unknown(probably left out of the ini), and 2 per damage type, Major and Minor. Moving the missile identifier in the INI to different subsections will color the missile according the the settings. For example Frozen Orb has 4 missiles; 
[IceMajor]
frozenorb=260
frozenorbnova=262
frozenorbexplode=263
...
[IceMinor]
frozenorbbolt=261
...

Or another example Blessed Hammers;
[MagicMajor]
blessedhammer=92
...
[MagicMinor]
blessedhammerex=640
...

There is a debug setting that turn a tooltip on for either "classified" missiles and otherMissiles/unknowns.

tthreeoh committed:
Mercs and Missiles

I wanted to show missiles and started by showing the mercs.

Missiles are turned off by default in settings.
There are two types of missiles, Player and Enemy, and both can be turned on/off separately.

As of right now I have 13 types of missiles each with a custom color in settings.

missiles.ini is used to process what missiles are included, ignored, unknown(probably left out of the ini), and 2 per damage type, Major and Minor. Moving the missile identifier in the INI to different subsections will color the missile according the the settings. For example Frozen Orb has 4 missiles;
[IceMajor]
frozenorb=260
frozenorbnova=262
frozenorbexplode=263
...
[IceMinor]
frozenorbbolt=261
...

Or another example Blessed Hammers;
[MagicMajor]
blessedhammer=92
...
[MagicMinor]
blessedhammerex=640
...

There is a debug setting that turn a tooltip on for either "classified" missiles and otherMissiles/unknowns.

tthreeoh committed:
typo

tthreeoh committed:
dynamic map location

made small changes to show map relative to game window

other minor changes

joffreybesos commented on src/MissilesRemoved.txt:
```diff
@@ -0,0 +1,123 @@
+blood1=18
```
With a separate text file that's read it would have to be bundled with every release.
You'll notice with other lists it's done as a switch statement in code. That way it gets compiled with the executable.

tthreeoh replied:
got ya!

joffreybesos commented on src/d2r-map.ahk:
```diff
@@ -1,22 +1,28 @@
-#SingleInstance, Force
+;#SingleInstance, Force
 #Persistent
+if not A_IsAdmin {
```
I've tried to avoid forcing to run as admin and leaving it to the user to run as admin if required. Is there a reason this is needed?

tthreeoh replied:
As far as I was aware the app HAS to run in admin so that it can read from a program that is in escalated memory.
Since I can't use the script without the escalation, I might as well add it. QOL, once less step. it doesnt't need to be there though  :p

joffreybesos commented on src/d2r-map.ahk:
```diff
@@ -107,6 +117,9 @@ unitHwnd1 := WinExist()
 offsetAttempts := 6
 ticktock := 0
 While 1 {
```
Should we still respect the 'leftMargin'' and 'topMargin' from the settings? So we would have `window left margin + left margin = actual margin`?

tthreeoh replied:
we could and probably should, though levelx and level y are pseudo margins in the mean time as they are persistent across sessions. I was gonna add a statement to set the map as docked or undocked.

joffreybesos commented on src/lib/automap.ahk:
```diff
@@ -0,0 +1,47 @@
+﻿class automap {
+
+}
```
Is this a class?

tthreeoh replied:
This is the beginning of a class. it's technically useless at this point but I wanted to leave it in. I'm working on consolidating the script functions into classes, nothing to see here... yet.

joffreybesos commented on src/memory/readGameMemory.ahk:
```diff
@@ -79,6 +80,19 @@ readGameMemory(d2rprocess, settings, playerOffset, ByRef gameMemoryData) {
+    missiles:=[]
+    ; PlayerMissiles
+    if (settings["showPlayerMissiles"]){
+        PlayerMissiles:=ReadUnits(d2rprocess, "0x21EA4F0", "Playermissiles")
```
I'd probably look at having a different way to specify this offset, probably just need (UnitHashTableOffset+ 1024*whatever)

tthreeoh replied:
Indeed! I'm learning a lot and so far the table hasn't moved! I'm still figuring out a lot but just as the automap class you noticed, in my readunits i have a switch case for different units. once I figure out the offest I will apply it but TBH ive never really messed with memory and offsets... like I get it but I've never really cared enough to do it. Now I do :p

joffreybesos commented on src/memory/readUnits.ahk:
```diff
@@ -0,0 +1,138 @@
+    return Matchlist
+}
+buildUnitList(catagory){
+    IniRead, %catagory%, missiles.ini, %catagory%
```
This is going to execute for each frame update. Do we want this to happen at startup instead?

tthreeoh replied:
I'm glad you saw this. This is why Im going to consolidate the script because... AHK. This actually wont update every frame, It will only run once. at the part it was included. AHK literally includes the code AT that location. 

AHK first, reads the script and the first return found not in a function or statement or hotkey ends the autoexecute portion of the script. so everything that we are including is technically ONLY getting executed once. The functions are "loaded into memory" and if the file contains code it will execute if it is within the auto execute. This means EVERY script config at the begining of the included files are executed and the begining of the first run. AHK runs by default so many lines, and then "takes a break" which we can configure using setbatchlines.

things like this only need to be in the script once unless you want to change it. but if you change it you'll know that you wanted to and probably make a statement to change it back after the needed use

#NoEnv
#MaxHotkeysPerInterval 99000000
#HotkeyInterval 99000000
#KeyHistory 0
ListLines Off
Process, Priority, , A
SetBatchLines, -1
SetKeyDelay, -1, -1
SetMouseDelay, -1
SetDefaultMouseSpeed, 0
SetWinDelay, -1
SetControlDelay, -1
SendMode Input

Since the Include is before your While loop, it is only run once and never touched again... which makes me want to put in a hotkey to turn different LOOT filters on dynamically  💡 i need to makea todo list....

joffreybesos reviewed (changes requested):
Hey I've left some comments. Probably needs to be tested as a standalone executable.

tthreeoh committed:
made changes

docked map with user margins

readsettings optimization for multiple

instances, and other fixes

missile.ini fileinstall

updated help screen to dock to game windows
